### DISCUSSION TITLE
Usage for filtering

### DISCUSSION THREAD

**Discussion by bodgerbarnett** [florimondmanca/djangorestframework-api-key#202] [Q&A]
I am developing a REST API with DRF for an events site which serves multiple domains. So there's a bunch of events that are particular to SiteA and a load more that are tied to SiteB.

Initially, I was planning to leave the API totally open (for reading) and allow filtering based on the site (something like /events/?site=sitea.com/) but then it occurred to me that anyone can then access every single event in my system which I don't really want.

I'm writing a frontend for it using Nuxt and I wondered about using this library to allow the event filtering. The idea would be that each Site would have it's own API_KEY and the Nuxt frontend would send this along with the HTTP requests to obtain the events. That API_KEY would then be tied to a particular Site in the database and that could be used to only return the appropriate events.

Before I go off down this road, is that a valid usage of this library? It sounds like it is but I just wanted to check that this all makes sense as I'm fairly new to DRF and this sort of thing.

Thanks in advance!

**Comment by florimondmanca** [comment]
Hi @bodgerbarnett. This hinted me at activating discussions on this repo. :)

From afar, it does seem like an appropriate use case.

E.g. a [custom API key setup](https://florimondmanca.github.io/djangorestframework-api-key/guide/#customization) that would link API keys to a `Site` (from Django's sites framework).

```python
class SiteAPIKey(AbstractAPIKey):
    site = models.ForeignKey(Site, on_delete=models.CASCADE)
```

Then you can protect ressources on each behind an API key tied to that domain. The rest of your "events" code would also think in terms of sites and domains, allowing you to filter events once permissions passed.

There's probably some fiddling required to carry out to find an appropriate implementation, though.

I think the key part is a [custom `.objects` manager](https://florimondmanca.github.io/djangorestframework-api-key/guide/#managers) that filters by site in `.get_usable_keys()`: `return super().get_usable_keys().filter(site=Site.objects.get_current())`

You'll probably also need a [custom permissions class](https://florimondmanca.github.io/djangorestframework-api-key/guide/#permission-classes) to link it to `SiteAPIKey`. The default permission class searches in `<model>.objects.get_usable_keys()` for valid keys, hence why you need to override that in a custom manager.

You may also want to consider what the [admin](https://florimondmanca.github.io/djangorestframework-api-key/guide/#admin-panel) should look like.

Feel free to look at the actual code to help yourself out, or ask if you need some help.

**Reply by bodgerbarnett** [reply]
Thanks - I guess I'm trying to head towards something similar to how Stripe do it. They obviously have tons of payments etc. in their system but you can only see the ones that are tied to your project via the API key you supply.

My system will be a lot smaller (we only have 2 sites at the moment - both set up by me, manually, using the Django Sites Framework) so when it comes to dishing out API keys, I can handle that myself, by hand, if need be. So the admin side of things might not be a problem.

The only thing I'm not sure about here is when you mention the custom permissions class. Are you talking about cases where someone other than the site it's tied to gets hold of the API key and uses it to access those events?

**Reply by florimondmanca** [reply]
> So the admin side of things might not be a problem.

I was mainly pointing out that if you end up with a custom API key model (linked to a `Site`), you'll probably need to register it in the admin. The default API key admin (for the default `APIKey` model) has a handy workflow for showing API keys on create.

> The only thing I'm not sure about here is when you mention the custom permissions class. Are you talking about cases where someone other than the site it's tied to gets hold of the API key and uses it to access those events?

Yes, I guess so. If I'm correct, to protect resources behind having an appropriate site API key, you'll need to apply a permission class on certain views. The default `HasAPIKey` permission class is tied to the default `APIKey` model. So you'll need to write your own that's tied to `SiteAPIKey` instead, and then use that on views.  This solves the technical part of actually searching the given API key against the `SiteAPIKey`'s. Then, the `.get_usable_keys()` bit would aim at preventing someone who might have API key for site B to be granted permission to access events for site A. That would only be an additional security/filtering measure since I suppose you'd do the actual filtering on `Event` by site anyway. Hope that makes sense and I understood your setup correctly!

**Reply by bodgerbarnett** [reply]
OK, sounds like I need to have a play with it then. I think I assumed that once the API key part was built, I could just use filtering to only return events that were connected to the site that the API key referenced. Then there would be no need for custom permissions. But as I said, I'm fairly new to this world so just trying to get my head round it all!

**Reply by marcosgeorgopoulos** [reply]
Hi @bodgerbarnett, I've just stumbled upon this library and had the exact same question as you. The example code with 'Heroes' let you create a HeroAPIKey that is mapped to a 'Hero' but then the code doesn't appear in anyway to restrict that Key to only viewing that 'Heroes' data.  I'm not sure of the point of the relationship? 

Did you ever get this working and did you take the approach suggested above?  and if so would you be willing to share some sample code?  

Many thanks,

**Reply by bodgerbarnett** [reply]
Hi @marcosgeorgopoulos,

I haven't got the whole system working properly yet but, so far, I've added a new permission called (using the example code) HasHeroAPIKey, like this:-

```
from rest_framework_api_key.permissions import BaseHasAPIKey

from .models import HeroAPIKey


class HasHeroAPIKey(BaseHasAPIKey):
    model = HeroAPIKey
```

which I use to ensure that *some* sort of API key has been provided by the client.

Then I've got a custom manager on the HeroAPIKey class, as follows:-

```
from rest_framework_api_key.models import APIKeyManager


class HeroAPIKeyManager(APIKeyManager):
    def get_from_request(self, request):
        from .permissions import HasHeroAPIKey

        key = HasHeroAPIKey().get_key(request)

        if key:
            return self.get_from_key(key)
```

which enables me to use the provided key to find the Hero it's connected to and then filter by that, like this:-

```
def get_queryset(self):
    hero_api_key = HeroAPIKey.objects.get_from_request(self.request)

    return super().get_queryset().filter(hero=hero_api_key.hero)
```

Hopefully that should get you started. I feel like the manager method needs some work as it doesn't look ideal but it's a start.
